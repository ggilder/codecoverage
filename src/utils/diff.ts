interface FileDiff {
  filename: string
  addedLines: number[]
  deletedLines: number[]
}

export function parseGitDiff(diffOutput: string): FileDiff[] {
  const fileDiffs: FileDiff[] = []
  const lines = diffOutput.split('\n')

  let currentFileDiff: FileDiff | undefined
  let currentAddedLines: number[] = []
  let currentDeletedLines: number[] = []
  let seenHeaderLine = false
  let deletionCurrentLineNumber = 0
  let additionCurrentLineNumber = 0

  for (const line of lines) {
    if (line.startsWith('diff --git')) {
      // New file diff starts
      if (currentFileDiff) {
        currentFileDiff.addedLines = currentAddedLines
        currentFileDiff.deletedLines = currentDeletedLines
        fileDiffs.push(currentFileDiff)
      }

      currentFileDiff = {
        filename: getFilenameFromDiffHeader(line),
        addedLines: [],
        deletedLines: []
      }
      currentAddedLines = []
      currentDeletedLines = []
      seenHeaderLine = false
    } else if (line.startsWith('@@')) {
      // Header line
      seenHeaderLine = true
      const lineInfo = getLineInfoFromHeaderLine(line)
      deletionCurrentLineNumber = lineInfo.deletionStartingLineNumber
      additionCurrentLineNumber = lineInfo.additionStartingLineNumber
    } else if (line.startsWith('+') && seenHeaderLine) {
      // Added line
      currentAddedLines.push(additionCurrentLineNumber)
      additionCurrentLineNumber++
    } else if (line.startsWith('-') && seenHeaderLine) {
      // Deleted line
      currentDeletedLines.push(deletionCurrentLineNumber)
      deletionCurrentLineNumber++
    } else if (seenHeaderLine) {
      // Context line
      deletionCurrentLineNumber++
      additionCurrentLineNumber++
    }
  }

  // Add the last file diff
  if (currentFileDiff) {
    currentFileDiff.addedLines = currentAddedLines
    currentFileDiff.deletedLines = currentDeletedLines
    fileDiffs.push(currentFileDiff)
  }

  return fileDiffs
}

function getFilenameFromDiffHeader(header: string): string {
  // Extract the filename from the diff header
  const startIndex = header.indexOf(' a/') + 3
  const endIndex = header.indexOf(' b/', startIndex)
  const filename = header.substring(startIndex, endIndex)
  return filename
}

function getLineInfoFromHeaderLine(line: string): {
  deletionStartingLineNumber: number
  additionStartingLineNumber: number
} {
  // Extract the starting line numbers for each side of the diff
  const matches = line.match(/-(\d+),?(\d+)? \+(\d+),?(\d+)? @@/)
  if (matches && matches.length === 5) {
    const deletionStartingLineNumber = parseInt(matches[1], 10)
    const additionStartingLineNumber = parseInt(matches[3], 10)
    return {deletionStartingLineNumber, additionStartingLineNumber}
  }
  return {deletionStartingLineNumber: 0, additionStartingLineNumber: 0}
}
