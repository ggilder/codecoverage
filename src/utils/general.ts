export function filterCoverageByFile(coverage: CoverageParsed): CoverageFile[] {
  return coverage.map(item => ({
    fileName: item.file,
    missingLineNumbers: item?.lines?.details
      .filter(line => line.hit === 0)
      .map(line => line.line)
  }))
}

export function coalesceLineNumbers(lines: number[]): LineRange[] {
  const ranges: LineRange[] = []
  let rstart
  let rend
  for (let i = 0; i < lines.length; i++) {
    rstart = lines[i]
    rend = rstart
    while (lines[i + 1] - lines[i] === 1) {
      rend = lines[i + 1]
      i++
    }
    ranges.push({start_line: rstart, end_line: rend})
  }
  return ranges
}

export function intersectLineRanges(
  rangesA,
  rangesB: LineRange[]
): LineRange[] {
  const outRanges: LineRange[] = []
  for (const bRange of rangesB) {
    const aRangeIntersects = rangesA.filter(aRange => {
      return (
        (bRange.start_line >= aRange.start_line &&
          bRange.start_line <= aRange.end_line) ||
        (bRange.end_line >= aRange.start_line &&
          bRange.end_line <= aRange.end_line) ||
        (aRange.start_line >= bRange.start_line &&
          aRange.start_line <= bRange.end_line) ||
        (aRange.end_line >= bRange.start_line &&
          aRange.end_line <= bRange.end_line)
      )
    })
    for (const aRange of aRangeIntersects) {
      outRanges.push({
        start_line: Math.max(aRange.start_line, bRange.start_line),
        end_line: Math.min(aRange.end_line, bRange.end_line)
      })
    }
  }
  return outRanges
}

export type CoverageParsed = {
  file: string
  title: string
  lines: {
    found: number
    hit: number
    details: {
      line: number
      hit: number
      name: string
    }[]
  }
}[]

export type CoverageFile = {
  fileName: string
  missingLineNumbers: number[]
}

export type LineRange = {
  start_line: number
  end_line: number
}
