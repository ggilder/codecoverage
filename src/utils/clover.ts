import * as clover from '@cvrg-report/clover-json'
import * as fs from 'fs'
import * as path from 'path'
import {CoverageParsed} from './general'

export async function parseClover(
  coveragePath: string,
  workspacePath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Clover XML path provided')
  }
  if (!workspacePath) {
    throw Error('No workspace path provided')
  }
  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  const parsed = await clover.parseContent(fileRaw)

  for (const entry of parsed) {
    entry.file = path.relative(workspacePath, entry.path || entry.file)
  }

  return parsed
}
