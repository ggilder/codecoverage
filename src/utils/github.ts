import * as core from '@actions/core'
import * as github from '@actions/github'
import {Octokit} from 'octokit'
import {CoverageFile} from './general'
import * as path from 'path'

export class GithubUtil {
  private client: Octokit

  constructor(token: string) {
    if (!token) {
      throw new Error('GITHUB_TOKEN is missing')
    }
    this.client = new Octokit({auth: token})
  }

  getPullRequestRef(): string {
    const pullRequest = github.context.payload.pull_request
    return pullRequest
      ? pullRequest.head.ref
      : github.context.ref.replace('refs/heads/', '')
  }

  /**
   * https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-pull-requests-files
   **/
  async getPullRequestFiles(): Promise<PullRequestFiles> {
    const pull_number = github.context.issue.number
    const response = await this.client.rest.pulls.listFiles({
      ...github.context.repo,
      pull_number
    })
    core.info(`Pull Request Files Length: ${response.data.length}`)

    return this.parsePullRequestFiles(response.data);
  }

  /**
   * https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#create-a-check-run
   * https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#update-a-check-run
   */
  async annotate(input: InputAnnotateParams): Promise<number> {
    if (input.annotations.length == 0) {
      return 0;
    }
    // github API lets you post 50 annotations at a time
    const chunkSize = 50;
    let chunks: Annotations[][] = [];
    for (let i = 0; i < input.annotations.length; i += chunkSize) {
      chunks.push(input.annotations.slice(i, i + chunkSize));
    }
    let lastResponseStatus = 0;
    let checkId;
    for (let i = 0; i < chunks.length; i++) {
      let status = 'in_progress';
      let conclusion = '';
      if (i == chunks.length - 1) {
        status = 'completed';
        conclusion = 'success';
      }
      let response;
      if (i == 0) {
        response = await this.client.rest.checks.create({
          ...github.context.repo,
          name: 'Annotate',
          head_sha: input.referenceCommitHash,
          status,
          ...(conclusion && {conclusion}),
          output: {
            title: 'Coverage Tool',
            summary: 'Missing Coverage',
            annotations: chunks[i]
          }
        })
        checkId = response.data.id;
      } else {
        response = await this.client.rest.checks.update({
          ...github.context.repo,
          name: 'Annotate',
          head_sha: input.referenceCommitHash,
          check_run_id: checkId,
          status,
          ...(conclusion && {conclusion}),
          output: {
            title: 'Coverage Tool',
            summary: 'Missing Coverage',
            annotations: chunks[i]
          }
        })
      }
      core.info(response.data.output.annotations_url);
      lastResponseStatus = response.status;
    }
    return lastResponseStatus;
  }

  buildAnnotations(
    coverageFiles: CoverageFile[],
    pullRequestFiles: PullRequestFiles,
    workspacePath: string
  ): Annotations[] {
    const annotations: Annotations[] = []
    for (const current of coverageFiles) {
      const relPath = path.relative(workspacePath, current.fileName)
      // Only annotate relevant files
      const prFileRanges = pullRequestFiles[relPath];
      if (prFileRanges) {
        const uncoveredRanges = this.coalesceLineNumbers(current.missingLineNumbers);
        // Only annotate relevant line ranges
        for (const uRange of uncoveredRanges) {
          const ok = prFileRanges.find(pRange => {
            return (uRange.start_line >= pRange.start_line && uRange.start_line <= pRange.end_line) ||
              (uRange.end_line >= pRange.start_line && uRange.end_line <= pRange.end_line)
          });
          // uncovered range overlaps a modified range in the PR
          if (ok) {
            annotations.push({
              path: relPath,
              start_line: uRange.start_line,
              end_line: uRange.end_line,
              start_column: 1,
              end_column: 1,
              annotation_level: 'warning',
              message: 'This line is not covered by a test'
            })
          }
        }
      }
    }
    core.info(`Annotation count: ${annotations.length}`)
    return annotations
  }

  parsePullRequestFiles(data): PullRequestFiles {
    const files: PullRequestFiles = {};
    for (const item of data) {
      files[item.filename] = this.parsePatchRanges(item.patch);
    }
    core.info(`Filename count: ${files.length}`);
    return files;
  }

  parsePatchRanges(patch: string): LineRange[] {
    let ranges: LineRange[] = [];
    const pattern = /@@ \-\d+,\d+ \+(\d+),(\d+) /g;

    let match;
    while (match = pattern.exec(patch)) {
      const start_line = parseInt(match[1], 10);
      const end_line = parseInt(match[2], 10) + start_line;
      const range: LineRange = { start_line, end_line };
      ranges.push(range);
    }

    return ranges;
  }

  coalesceLineNumbers(lines: number[]): LineRange[] {
    const ranges: LineRange[] = [];
    let rstart, rend;
    for (let i = 0; i < lines.length; i++) {
      rstart = lines[i];
      rend = rstart;
      while (lines[i + 1] - lines[i] === 1) {
        rend = lines[i + 1];
        i++;
      }
      ranges.push({ start_line: rstart, end_line: rend });
    }
    return ranges;
  }
}

type InputAnnotateParams = {
  referenceCommitHash: string
  annotations: Annotations[]
}

type Annotations = {
  path: string
  start_line: number
  end_line: number
  start_column: number
  end_column: number
  annotation_level: string
  message: string
}

type LineRange = {
  start_line: number
  end_line: number
}

type PullRequestFiles = {
  [key: string]: LineRange[]
}
