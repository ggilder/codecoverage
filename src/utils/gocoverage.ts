import * as fs from 'fs'
import * as gocov from 'golang-cover-parse'
import * as path from 'path'
import * as readline from 'readline'
import {CoverageParsed} from './general'

export async function parseGoCoverage(
  coveragePath: string,
  goModPath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Go coverage path provided')
  }

  if (!goModPath) {
    throw Error('No Go module path provided')
  }

  const goModule = await parseGoModFile(goModPath)

  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  return new Promise((resolve, reject) => {
    gocov.parseContent(fileRaw, (err, result) => {
      if (err === null) {
        filterModulePaths(result, goModule)
        resolve(result)
      }
      reject(err)
    })
  })
}

function filterModulePaths(entries, moduleName: string): void {
  for (const entry of entries) {
    entry.file = path.relative(moduleName, entry.file)
  }
}

async function parseGoModFile(filePath: string): Promise<string> {
  const fileStream = fs.createReadStream(filePath)
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  })
  // Note: we use the crlfDelay option to recognize all instances of CR LF
  // ('\r\n') in input.txt as a single line break.

  for await (const line of rl) {
    if (line.startsWith('module ')) {
      return line.slice(7)
    }
  }

  /* istanbul ignore next */
  return ''
}
