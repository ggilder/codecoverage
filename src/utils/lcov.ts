import * as NodeUtil from 'util'
import * as fs from 'fs'
import * as path from 'path'
import {CoverageParsed} from './general'
import parser from 'lcov-parse'

export async function parseLCov(
  lcovPath: string,
  workspacePath: string
): Promise<CoverageParsed> {
  if (!lcovPath) {
    throw Error('No LCov path provided')
  }

  const parserPromise = NodeUtil.promisify(parser)
  const fileRaw = fs.readFileSync(lcovPath, 'utf8')
  const parsed = (await parserPromise(fileRaw)) as CoverageParsed

  for (const entry of parsed) {
    entry.file = path.relative(workspacePath, entry.file)
  }

  return parsed
}
