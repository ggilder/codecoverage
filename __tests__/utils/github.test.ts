import {test} from '@jest/globals'
import {GithubUtil} from '../../src/utils/github'

test('github init successfully', async function () {
  const githubUtil = new GithubUtil('1234')
  expect(githubUtil).toBeInstanceOf(GithubUtil)
})

test('github init to throw error', function () {
  expect(() => new GithubUtil('')).toThrowError('GITHUB_TOKEN is missing')
})

test('build annotations', function () {
  const githubUtil = new GithubUtil('1234')

  const prFiles = {
    'file1.txt': [
      {end_line: 139, start_line: 132},
      {end_line: 1007, start_line: 1000}
    ],
    'test/dir/file1.txt': [{end_line: 45, start_line: 22}]
  }

  const coverageFiles = [
    {fileName: 'unchanged.txt', missingLineNumbers: [1, 2, 3]},
    {
      fileName: 'file1.txt',
      missingLineNumbers: [1, 2, 3, 132, 134, 135, 136, 1007, 1008]
    },
    {
      fileName: 'test/dir/file1.txt',
      missingLineNumbers: [20, 21, 22]
    }
  ]

  const annotations = githubUtil.buildAnnotations(coverageFiles, prFiles)

  expect(annotations).toEqual([
    {
      path: 'file1.txt',
      start_line: 132,
      end_line: 132,
      annotation_level: 'warning',
      message: 'This line is not covered by a test'
    },
    {
      path: 'file1.txt',
      start_line: 134,
      end_line: 136,
      annotation_level: 'warning',
      message: 'These lines are not covered by a test'
    },
    {
      path: 'file1.txt',
      start_line: 1007,
      end_line: 1007,
      annotation_level: 'warning',
      message: 'This line is not covered by a test'
    },
    {
      path: 'test/dir/file1.txt',
      start_line: 22,
      end_line: 22,
      annotation_level: 'warning',
      message: 'This line is not covered by a test'
    }
  ])

  const consoleMock = jest.spyOn(console, 'log').mockImplementation()

  githubUtil.annotate(annotations)
  expect(console.log).toHaveBeenCalledTimes(4)
  expect(console.log).toHaveBeenCalledWith(
    '::warning file=file1.txt,line=132,endLine=132::This line is not covered by a test'
  )
  expect(console.log).toHaveBeenCalledWith(
    '::warning file=file1.txt,line=134,endLine=136::These lines are not covered by a test'
  )
  expect(console.log).toHaveBeenCalledWith(
    '::warning file=file1.txt,line=1007,endLine=1007::This line is not covered by a test'
  )
  expect(console.log).toHaveBeenCalledWith(
    '::warning file=test/dir/file1.txt,line=22,endLine=22::This line is not covered by a test'
  )

  consoleMock.mockRestore()
})

// @todo test for rest of github class
