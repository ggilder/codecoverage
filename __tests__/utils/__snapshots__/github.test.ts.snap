// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse pull request files 1`] = `
Object {
  "file1.txt": Array [
    Object {
      "end_line": 139,
      "start_line": 132,
    },
    Object {
      "end_line": 1007,
      "start_line": 1000,
    },
  ],
  "test/dir/file1.txt": Array [
    Object {
      "end_line": 45,
      "start_line": 22,
    },
  ],
}
`;
